### DISCUSSION TITLE
Source code for the toy browser?

### DISCUSSION THREAD

**Discussion by gabrielpetersson** [browserengineering/book#694] [Ideas]
Would be awesome to be able to follow the source code through every chapter, so you have an overview of everything that has been written at any chapter. I'm just reading it through & would be awesome to have the code to refer to, as it gets hard to keep all concepts in your head when you are hetting further into the book!

awesome work

**Comment by pavpanchekha** [comment]
Yep—you can find it in the Github repo under `src/labXXX.py`. The `XXX` refers to the chapter number. These versions still include imports from prior chapters, but if you don't want that you can try running `make src/labXXX.full.py`, which will run an inliner we wrote to inline all the dependencies.

**Reply by gabrielpetersson** [reply]
Oh awesome! Thanks

**Comment by Paganiniana** [comment]
Quick question on this one (maybe you'll know since you've been around this a bit longer, @gabrielpetersson). I'm trying to sync some differences between my TS project and the official Chapter 5 source. The source's `layout` method is calling _self.recurse_ , but I can't figure out where the definition is. It's included in the file for Chapter 4's **Layout** class, though. Is this something that needs to be copied over?

**Reply by Paganiniana** [reply]
(^) Oops, sorry, my eyes completely glossed over the `@wbetools.patch` decorator. 🙄 Never mind! On that note, though, I wonder if it wouldn't be kind of cool to have some browsable "state at chapter N" — a visually attractive version of that build you mentioned, @pavpanchekha, that's included at the end of each chapter. I understand the point of _Browse Book_ is to build it yourself (I'm a through-and-through Constructionist, so I'm totally on board with that). It might be cool for people who haven't spent time in large-ish code bases to see the size of the project build, iteratively, even if they aren't building along with it. Just spitballing!

**Reply by pavpanchekha** [reply]
Yeah, it would. As you've probably seen, you can get a "complete" version by running `make src/labN.full.py` and opening that file. Making that pretty would be nice but probably something to work on after we finish planned changes and refactors.

**Reply by Paganiniana** [reply]
For sure. There's always something you _could_ add. You guys are doing such a good job prioritizing and making progress though. 👏
